File: js-python-educative-challenges/two-pointers/palindrome_string_chance.js
// Write a function that takes a string as input and checks whether it can be
//  a valid palindrome by removing at most one character from it.

/**
 * @param {string} s
 * @returns {bool}
 */

function isPalindromeSlice(s) {
  if (s.length < 1) return false;
  let [left, right] = [0, s.length - 1]
  while (left < right) {
    if (s[left] !== s[right]) return false;
    left++;
    right--;
  }
  return true;
}


/**
 * @param {string} s
 * @returns {bool}
 */

function couldBePalindrome(s) {
  if (s.length < 1) return false;
  let [left, right, mismatch] = [0, s.length - 1, 0]
  while (left < right) {
    if (s[left] === s[right]) {
      left++;
      right--;
    } else {
      mismatch++;
      if (mismatch > 1) return false;
      if (isPalindromeSlice(s.slice(left, right))) return true;
      if (isPalindromeSlice(s.slice(left+1, right+1))) return true;
    }
  }
  return true;
}


// main

const s = 'abcdedadedecba'  // abc dedaded (e) cba
const res = couldBePalindrome(s)
console.log(res)

File: js-python-educative-challenges/two-pointers/palindrome_string_chance.py
"""Write a function that takes a string as input and checks whether it can be
 a valid palindrome by removing at most one character from it."""

def is_palindrome_slice(s: str) -> bool:
    print(s)
    if len(s) < 1:
        return False

    left, right = 0, len(s) - 1

    while left < right:
        if s[left] != s[right]:
            return False
        left += 1
        right -= 1
    return True


def could_it_be_palindrome(s: str) -> bool:
    if len(s) < 1:
        return False

    left, right, dismatch = 0, len(s) - 1, 0

    while left < right:
        if s[left] == s[right]:
            left += 1
            right -= 1

        else:
            dismatch += 1
            if dismatch > 1:
                return False

            if is_palindrome_slice(s[left:right]):
                return True

            if is_palindrome_slice(s[left+1:right+1]):
                return True
    return True


if __name__ == '__main__':
    # s = 'madame'
    s = 'abcdedadedecba'  # abc dedaded (e) cba
    res = could_it_be_palindrome(s)
    print(res)
